### PR TITLE
Add gradient to BiasAddGrad op.

### LINKED ISSUE #4174: tf.nn.bias_add does not support multiple derivatives (using with contrib.learn)
Apparently, `tf.nn.bias_add` does not support differentiating twice: I created a neural network with `tf.contrib.layers`, and attempting to compute a Hessian matrix (actually Hessian-vector product, based on https://en.wikipedia.org/wiki/Hessian_automatic_differentiation) yields a `LookupError` saying that `BiasAddGrad` has no gradient.

From what I understand, `bias_add` is basically `add` with support for different types.

Is there any way to allow `bias_add` to be differentiated multiple times? Perhaps at the very least we could make the use of `bias_add` optional in `layers`.

### PR DESCRIPTION
**Pull request by admcrae**
A rather inelegant solution to fix #4174.

### CODE REVIEW

File: tensorflow/python/ops/nn_grad.py [hunk 1]
```diff
@@ -203,6 +204,27 @@ def _BiasAddGrad(op, received_grad):
...
+    is the gradient of the bias term in BiasAdd)
+  """
+  
+  try:
+    data_format = op.get_attr("data_format")
+  except ValueError:
+    data_format = None
+    
+  zeros = gen_array_ops._zeros_like(op.inputs[0])
+  return gen_nn_ops._bias_add(zeros, received_grad, data_format=data_format)
```
Comment (zheng-xq) [inline tensorflow/python/ops/nn_grad.py]: In cases where the shapes are known, I think a broadcasting op such as tf.tile will be better. Since it avoids allocating the zero input array and add the unnecessary zeros. However, tf.tile in its current form may be slower. So a bit benchmarking is needed for that path. I'll leave it up to you whether you want to try that. 

Also we need to add a test using compute_gradient_errort to make sure this passes the gradient checker.

Comment (admcrae) [inline tensorflow/python/ops/nn_grad.py]: I don't see a file in tensorflow/python/ops for testing nn gradient ops. Is it somewhere else? Otherwise, should I create a new nn_grad_test file or just use the current nn_test file?

Comment (zheng-xq) [inline tensorflow/python/ops/nn_grad.py]: Please use bias_op_test. Thanks.

File: tensorflow/python/kernel_tests/bias_op_test.py [hunk 2]
```diff
@@ -128,7 +129,16 @@ def _testGradient(self, np_input, bias, dtype, data_format, use_gpu):
           input_tensor, np_input.shape, output_tensor, np_input.shape)
       bias_jacob_t, bias_jacob_n = tf.test.compute_gradient(
           bias_tensor, bias.shape, output_tensor, np_input.shape)
-
+         
+      # Test gradient of BiasAddGrad
+      bias_add_grad = gen_nn_ops.bias_add_grad(out_backprop=output_tensor,
```
Comment (zheng-xq) [inline tensorflow/python/kernel_tests/bias_op_test.py]: Instead of calling the underlying function, maybe calling "tf.Gradient(loss, bias)". So even if the underlying implementation changes, the test might still be valid.

File: tensorflow/python/ops/nn_grad.py [hunk 3]
```diff
@@ -203,6 +203,48 @@ def _BiasAddGrad(op, received_grad):
...
+    data_format = op.get_attr("data_format")
+  except ValueError:
+    data_format = None
+    
+  #zeros = array_ops.zeros_like(op.inputs[0])
+  #return gen_nn_ops._bias_add(zeros, received_grad, data_format=data_format)
+  
+  shape = array_ops.shape(op.inputs[0])
+  rank = array_ops.rank(op.inputs[0])
+  bias_shape = array_ops.shape(received_grad)
```
Comment (zheng-xq) [inline tensorflow/python/ops/nn_grad.py]: If the shape is not fully defined at this point, this adds an unnecessary dependency at runtime. So it is better to do this when the shape is known, and do your previous bias_add when it is not. For example:

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/array_ops.py#L154

If possible, please also include a benchmark result to show this is not actually slower for non-trivially large shape. tf.tile may have a few performance problems on GPU, and it is on our roadmap for improvement.

Comment (admcrae) [inline tensorflow/python/ops/nn_grad.py]: Some crude benchmarks:
[benchmark.zip](https://github.com/tensorflow/tensorflow/files/479926/benchmark.zip)

The benchmarks were done on a workstation with an i7-965 (yes, we laughed too when we got it) and a GTX 1080.

The results are a bit odd. Using `bias_add` with `zeros_like` seems to give a negligible to moderate performance advantage in most cases, but it is 2-5x slower on the CPU with statically determined shapes (it looks like something is not being optimized properly here). I did notice that on the GPU, using `bias_add` seemed to result in somewhat higher GPU usage overall.

I'm inclined to stay with `tf.tile` for now, since (1) it is a more direct solution (using `bias_add` seems extremely hacky), (2) its performance is more consistent, and (3) you mentioned that the performance of `tf.tile` might be improved in the future.

Of course, the optimal solution would be to write another kernel for this operation (and then we could finally end the chain of gradient ops, since this operation's gradient is identical to `bias_add_grad`), but I'm not sure if that is worthwhile for something that most machine learning people aren't going to need.

File: tensorflow/python/ops/nn_grad.py [hunk 4]
```diff
@@ -203,6 +203,43 @@ def _BiasAddGrad(op, received_grad):
...
+    is the gradient of the bias term in BiasAdd)
+  """
+  
+  try:
+    data_format = op.get_attr("data_format")
+  except ValueError:
+    data_format = None
+  
+  shape = array_ops.shape(op.inputs[0])
+  rank = array_ops.rank(op.inputs[0])
```
Comment (zheng-xq) [inline tensorflow/python/ops/nn_grad.py]: Could you assert that rank is 4 for now? For conv3d and 5D tensors from videos, we will have to make adjustment. However, there are plenty of other code that needs to be modified. So just a check would suffice for now.

**Review by zheng-xq — COMMENTED** [review]
I agree with the assessment on tf.tile is better for now. With a bit more nitpicking, I think we will be ready to go.

**Comment by martinwicke** [comment]
Changing assignment to @zheng-xq to reflect the reality of the review.

File: tensorflow/python/ops/nn_grad.py (hunk 4 above)
Comment (admcrae) [inline tensorflow/python/ops/nn_grad.py]: `contrib.layers` [uses `bias_add` even for fully-connected layers](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/layers/python/layers/layers.py#L857) (where the tensors are 2-D). I'd be hesitant to mess that up if anyone wanted to do higher-order gradients with that code (that's where I originally noticed this issue, although I'm not using it right now).

Comment (zheng-xq) [inline tensorflow/python/ops/nn_grad.py]: Thanks a lot for getting to that. It is indeed an important use case. Could you add a test case to cover that as well? We need to make sure future changes do not break that.

Comment (admcrae) [inline tensorflow/python/ops/nn_grad.py]: [`bias_op_test.py`](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/kernel_tests/bias_op_test.py) appears to have plenty of tests for ranks 2 and 3 but, curiously, no test for rank-4 tensors (for the bias_add op itself; it has gradient tests for ranks 2 and 4). Is this what you meant?

File: tensorflow/python/kernel_tests/bias_op_test.py [hunk 5]
```diff
@@ -128,7 +128,16 @@ def _testGradient(self, np_input, bias, dtype, data_format, use_gpu):
...
-
+         
+      # Test gradient of BiasAddGrad
+      bias_add_grad = tf.gradients(tf.nn.l2_loss(output_tensor),
+                                   bias_tensor)[0]
+      
+      grad_jacob_t, grad_jacob_n = tf.test.compute_gradient(
+          output_tensor, np_input.shape, bias_add_grad, bias.shape
+      )
+
```
Comment (zheng-xq) [inline tensorflow/python/kernel_tests/bias_op_test.py]: One less new line please.

File: tensorflow/python/kernel_tests/bias_op_test.py [hunk 6]
```diff
@@ -144,12 +153,22 @@ def _testGradient(self, np_input, bias, dtype, data_format, use_gpu):
             input_tensor, np_input.shape, output_tensor, np_input.shape)
         _, bias_jacob_n = tf.test.compute_gradient(
             bias_tensor, bias.shape, output_tensor, np_input.shape)
-
+
```
Comment (zheng-xq) [inline tensorflow/python/kernel_tests/bias_op_test.py]: Same here.

File: tensorflow/python/kernel_tests/bias_op_test.py [hunk 7]
```diff
@@ -144,12 +153,22 @@ def _testGradient(self, np_input, bias, dtype, data_format, use_gpu):
             input_tensor, np_input.shape, output_tensor, np_input.shape)
         _, bias_jacob_n = tf.test.compute_gradient(
             bias_tensor, bias.shape, output_tensor, np_input.shape)
-
+        
+        
+        bias_add_grad = tf.gradients(tf.nn.l2_loss(output_tensor),
+                                     bias_tensor)[0]
+
```
Comment (zheng-xq) [inline tensorflow/python/kernel_tests/bias_op_test.py]: And here.

**Review by zheng-xq — COMMENTED** [review]
Looks good, other than cosmetic suggestions.

**Comment by admcrae** [comment]
That better? Note I haven't made any changes to the tests since your previous comment--I wasn't entirely sure what you wanted there.

**Comment by zheng-xq** [comment]
LGTM.

@tensorflow-jenkins test this please.

**Review by martinwicke — COMMENTED** [review]
reusing zheng-xq's LGTM

**Review by martinwicke — APPROVED** [review]

**Comment by zheng-xq** [comment]
@martinwicke, could you confirm the mac test failure is unrelated? Thanks.

**Comment by martinwicke** [comment]
Yes, unrelated. Good to merge.

**Comment by martinwicke** [comment]
@jhseu feel free to use your powers and override the failed test.

File: tensorflow/python/kernel_tests/bias_op_test.py [hunk 8]
```diff
@@ -128,7 +128,13 @@ def _testGradient(self, np_input, bias, dtype, data_format, use_gpu):
           input_tensor, np_input.shape, output_tensor, np_input.shape)
       bias_jacob_t, bias_jacob_n = tf.test.compute_gradient(
           bias_tensor, bias.shape, output_tensor, np_input.shape)
-
+         
+      # Test gradient of BiasAddGrad
+      bias_add_grad = tf.gradients(tf.nn.l2_loss(output_tensor),
+                                   bias_tensor)[0]
+      grad_jacob_t, grad_jacob_n = tf.test.compute_gradient(
+          output_tensor, np_input.shape, bias_add_grad, bias.shape)
```
Comment (gunan) [inline tensorflow/python/kernel_tests/bias_op_test.py]: I think we have a python3 issue here. Our nightlies are failing on this test:
http://ci.tensorflow.org/view/Nightly/job/nightly-matrix-linux-gpu/TF_BUILD_IS_OPT=OPT,TF_BUILD_IS_PIP=NO_PIP,TF_BUILD_PYTHON_VERSION=PYTHON3,label=gpu-linux/247/consoleFull

I was also able to reproduce the problem locally using the following command:
$ bazel test tensorflow/python/kernel_tests:bias_op_test -c opt --config=cuda

Comment (admcrae) [inline tensorflow/python/kernel_tests/bias_op_test.py]: Is there anything I can do here? This appears to be limited to Python3 with a GPU, so it seems unlikely that this is due to my code.

Comment (admcrae) [inline tensorflow/python/kernel_tests/bias_op_test.py]: To clarify, right now the data_format of the bias op is being compared to a Python string literal. Looking at [another similar code snippet](https://github.com/tensorflow/tensorflow/blob/450e9386e0cc4b81b49adb9184859411f43ef259/tensorflow/python/framework/common_shapes.py#L185) suggests that should explicitly make it an ASCII string. I'll fix that shortly.

Comment (gunan) [inline tensorflow/python/kernel_tests/bias_op_test.py]: Sorry, I was reading through my email backlog, and just saw your update.
Thank you very much for looking into this!
